ateam_common/src/assignment.cpp

```diff
@@ -0,0 +1,288 @@
+}
+
+namespace internal
+{
+Eigen::MatrixXd squarize_matrix(const Eigen::MatrixXd & matrix)
+{
+  if (matrix.rows() == matrix.cols()) {
+    return matrix;
+  }
+
+  std::size_t new_row_col = std::max(matrix.rows(), matrix.cols());
+  Eigen::MatrixXd new_mat =
+    Eigen::MatrixXd::Constant(new_row_col, new_row_col, UNFILLED_LARGE_VALUE);
+
+  new_mat.block(0, 0, matrix.rows(), matrix.cols()) = matrix;
+
+  return new_mat;
+}
+
+bool has_unique_assignments(const CostMarkCovers & cmc)
+{
+  // The count of marked rows and columns should be the size of the matrix
+  int num_marked = 0;
+  for (int i = 0; i < cmc.cost_matrix.rows(); i++) {
+    num_marked += cmc.row_covers(i);
+    num_marked += cmc.col_covers(i);
+  }
+
+  return num_marked == cmc.cost_matrix.rows();
+}
+
+void apply_step_1(CostMarkCovers & cmc)
+{
+  Eigen::VectorXd min_coeff = cmc.cost_matrix.rowwise().minCoeff();
+  for (int i = 0; i < cmc.cost_matrix.rows(); i++) {
+    cmc.cost_matrix.row(i) -= Eigen::MatrixXd::Constant(1, cmc.cost_matrix.cols(), min_coeff(i));
+  }
+}
+
+void apply_step_2(CostMarkCovers & cmc)
+{
+  Eigen::VectorXd min_coeff = cmc.cost_matrix.colwise().minCoeff();
+  for (int i = 0; i < cmc.cost_matrix.cols(); i++) {
+    cmc.cost_matrix.col(i) -= Eigen::MatrixXd::Constant(cmc.cost_matrix.rows(), 1, min_coeff(i));
+  }
+}
+
+void apply_step_3(CostMarkCovers & cmc)
+{
+  cmc.row_covers = Eigen::VectorXi::Zero(cmc.cost_matrix.rows());
+  cmc.col_covers = Eigen::VectorXi::Zero(cmc.cost_matrix.cols());
+  cmc.mark_matrix = Eigen::MatrixXi::Zero(cmc.cost_matrix.rows(), cmc.cost_matrix.cols());
+
+  // Mark uncovered zeros one at a time
+  for (int i = 0; i < cmc.cost_matrix.rows(); i++) {
+    bool is_row_covered = cmc.row_covers(i) > 0;
+    if (is_row_covered) {
+      continue;
+    }
+    for (int j = 0; j < cmc.cost_matrix.cols(); j++) {
+      // Can't count zeros that are already covered
+      bool is_col_covered = cmc.col_covers(j) > 0;
+      if (is_col_covered) {
+        continue;
+      }
+
+      bool is_zero = cmc.cost_matrix(i, j) == 0;
+      if (is_zero) {
+        cmc.mark_matrix(i, j) = ZerosType::STARRED;
+        cmc.row_covers(i) = 1;
+        cmc.col_covers(j) = 1;
+        break;
+      }
+    }
+  }
+}
+
+void star_zero_cols(CostMarkCovers & cmc)
+{
+  cmc.col_covers = Eigen::VectorXi::Zero(cmc.mark_matrix.cols());
+  Eigen::VectorXi marked_col_coeff = cmc.mark_matrix.colwise().maxCoeff();
+  for (int j = 0; j < cmc.mark_matrix.cols(); j++) {
+    bool is_starred_zero = marked_col_coeff(j) == ZerosType::STARRED;
+    if (is_starred_zero) {
+      cmc.col_covers(j) = 1;
+    }
+  }
+}
+
+std::optional<Eigen::Vector2i> next_uncovered_zero(
+  const CostMarkCovers & cmc)
+{
+  for (int i = 0; i < cmc.cost_matrix.rows(); i++) {
+    for (int j = 0; j < cmc.cost_matrix.cols(); j++) {
+      bool is_covered_col = cmc.col_covers(j) == 1;
+      bool is_covered_row = cmc.row_covers(i) == 1;
+      bool is_zero = cmc.cost_matrix(i, j) == 0;
```

Comment:
You could continue to the next loop iteration after each condition here if they fail rather than checking all 3 at once. Not sure if that's more optimal in terms of speed or not.

Reply:
True true, pulled the `is_covered_row` continue into the outer loop, and separated the `is_covered_col` continue into a difference if statement